internal/pkg/coordinator/monitor.go

```diff
@@ -0,0 +1,369 @@
+	leaders := map[string]model.PolicyLeader{}
+	policies, err := dl.QueryLatestPolicies(ctx, m.bulker, dl.WithIndexName(m.policiesIndex))
+	if err != nil {
+		return err
+	}
+	if len(policies) > 0 {
+		ids := make([]string, len(policies))
+		for i, p := range policies {
+			ids[i] = p.PolicyId
+		}
+		leaders, err = dl.SearchPolicyLeaders(ctx, m.bulker, ids, dl.WithIndexName(m.leadersIndex))
+		if err != nil {
+			return err
+		}
+	}
+
+	// determine the policies that lead needs to be taken
+	lead := []model.Policy{}
+	now := time.Now().UTC()
+	for _, policy := range policies {
+		leader, ok := leaders[policy.PolicyId]
+		if !ok {
+			// new policy want to try to take leadership
+			lead = append(lead, policy)
+			continue
+		}
+		t, err := leader.Time()
+		if err != nil {
+			return err
+		}
+		if now.Sub(t) > m.leaderInterval || leader.Server.Id == m.agentMetadata.Id {
+			// policy needs a new leader or already leader
+			lead = append(lead, policy)
+		}
+	}
+
+	// take/keep leadership and start new coordinators
+	res := make(chan policyT)
+	for _, p := range lead {
+		pt, _ := m.policies[p.PolicyId]
+		pt.id = p.PolicyId
+		go func(p model.Policy, pt policyT) {
+			l := m.log.With().Str(dl.FieldPolicyId, pt.id).Logger()
+			err := dl.TakePolicyLeadership(ctx, m.bulker, pt.id, m.agentMetadata.Id, m.version, dl.WithIndexName(m.leadersIndex))
+			if err != nil {
+				l.Err(err).Msg("failed to take ownership")
+				if pt.cord != nil {
+					pt.canceller()
+					pt.canceller = nil
+					pt.cord = nil
+				}
+				res <- pt
+				return
+			}
+			if pt.cord == nil {
+				cord, err := m.factory(p)
+				if err != nil {
+					l.Err(err).Msg("failed to start coordinator")
+					_ = dl.ReleasePolicyLeadership(ctx, m.bulker, pt.id, m.agentMetadata.Id, m.leaderInterval, dl.WithIndexName(m.leadersIndex))
+					res <- pt
+					return
+				}
+
+				cordCtx, canceller := context.WithCancel(ctx)
+				go runCoordinator(cordCtx, cord, l, m.coordRestartDelay)
+				go runCoordinatorOutput(cordCtx, cord, m.bulker, l, m.policiesIndex)
+				pt.cord = cord
+				pt.canceller = canceller
+			}
+			res <- pt
+		}(p, pt)
+	}
+	for range lead {
+		r := <-res
+		if r.cord == nil {
+			// either failed to take leadership or lost leadership
+			delete(m.policies, r.id)
+		} else {
+			m.policies[r.id] = r
+		}
+	}
+	return nil
+}
+
+// releaseLeadership releases current leadership
+func (m *monitorT) releaseLeadership() {
+	var wg sync.WaitGroup
+	wg.Add(len(m.policies))
+	for _, pt := range m.policies {
+		go func(pt policyT) {
+			if pt.cord != nil {
+				pt.canceller()
+			}
+			// uses a background context, because the context for the
+			// monitor will be cancelled at this point in the code
+			err := dl.ReleasePolicyLeadership(context.Background(), m.bulker, pt.id, m.agentMetadata.Id, m.leaderInterval, dl.WithIndexName(m.leadersIndex))
+			if err != nil {
+				l := m.log.With().Str(dl.FieldPolicyId, pt.id).Logger()
+				l.Err(err).Msg("failed to release leadership")
```

Comment:
nit:
what's the convention for the log messages? 
was following with the capital letter start convention before in other places.
Just need to make it consistent one way or another.

Reply:
i am a big fan of lowercase log messages everywhere, think I only got the failed and start wrong on the monitor

Reply:
ok, will start moving to that standard then on the pieces that I touch